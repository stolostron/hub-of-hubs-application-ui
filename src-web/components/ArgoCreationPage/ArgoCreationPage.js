/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */
// Copyright (c) 2020 Red Hat, Inc.
// Copyright Contributors to the Open Cluster Management project
import React from 'react'
import { PageSection } from '@patternfly/react-core'
import resources from '../../../lib/shared/resources'
import msgs from '../../../nls/platform.properties'
import { Query } from 'react-apollo'
import { getArgoApplication } from '../../../lib/client/queries'
import PropTypes from 'prop-types'
import { withRouter } from 'react-router-dom'
import { connect } from 'react-redux'
import {
  createApplication,
  updateApplication,
  clearCreateStatus
} from '../../actions/application'
import { controlData } from './controlData/controlData'
import createTemplate from './templates/template.hbs'
import { updateSecondaryHeader } from '../../actions/common'
import { canCreateActionAllNamespaces } from '../../../lib/client/access-helper'
import config from '../../../lib/shared/config'
import _ from 'lodash'

// include monaco editor
import TemplateEditor from 'temptifly'
import 'temptifly/dist/styles.css'
import MonacoEditor from 'react-monaco-editor'
import 'monaco-editor/esm/vs/editor/editor.all.js'
import 'monaco-editor/esm/vs/editor/standalone/browser/quickOpen/quickCommand.js'
import 'monaco-editor/esm/vs/basic-languages/yaml/yaml.contribution.js'
import { global_BackgroundColor_dark_100 as editorBackground } from '@patternfly/react-tokens'
import { getApplicationResources } from './transformers/transform-data-to-resources'
import classNames from 'classnames'

if (window.monaco) {
  window.monaco.editor.defineTheme('console', {
    base: 'vs-dark',
    inherit: true,
    rules: [
      // avoid pf tokens for `rules` since tokens are opaque strings that might not be hex values
      { token: 'number', foreground: 'ace12e' },
      { token: 'type', foreground: '73bcf7' },
      { token: 'string', foreground: 'f0ab00' },
      { token: 'keyword', foreground: 'cbc0ff' }
    ],
    colors: {
      'editor.background': editorBackground.value,
      'editorGutter.background': '#292e34', // no pf token defined
      'editorLineNumber.activeForeground': '#fff',
      'editorLineNumber.foreground': '#f0f0f0'
    }
  })
}

window.MonacoEnvironment = {
  getWorkerUrl: function() {
    return `${config.contextPath}/editor.worker.js`
  }
}

const Portals = Object.freeze({
  editBtn: 'edit-button-portal-id',
  cancelBtn: 'cancel-button-portal-id',
  createBtn: 'create-button-portal-id'
})

resources(() => {
  require('./style.scss')
})

class ArgoCreationPage extends React.Component {
  static propTypes = {
    cleanReqStatus: PropTypes.func,
    handleCreateApplication: PropTypes.func,
    handleUpdateApplication: PropTypes.func,
    history: PropTypes.object,
    location: PropTypes.object,
    match: PropTypes.object,
    mutateErrorMsgs: PropTypes.array,
    mutateStatus: PropTypes.string,
    secondaryHeaderProps: PropTypes.object,
    updateSecondaryHeader: PropTypes.func
  };

  constructor(props) {
    super(props)
    this.state = {
      hasPermissions: false
    }
    this.getBreadcrumbs = this.getBreadcrumbs.bind(this)
  }

  getEditApplication() {
    const { match: { params } } = this.props
    if (params.name && params.namespace) {
      return {
        selectedAppName: params.name,
        selectedAppNamespace: params.namespace
      }
    }
    return null
  }

  getBreadcrumbs() {
    const { location } = this.props,
          urlSegments = location.pathname.split('/')
    return [
      {
        label: msgs.get('resource.applications', this.context.locale),
        url: urlSegments.slice(0, Math.min(3, urlSegments.length)).join('/')
      }
    ]
  }

  componentDidMount() {
    const { secondaryHeaderProps = {}, cleanReqStatus } = this.props
    const { selectedAppName } = this.getEditApplication() || {}
    const { locale } = this.context
    if (cleanReqStatus) {
      this.props.cleanReqStatus()
    }
    const portals = [
      {
        id: 'edit-button-portal-id',
        kind: 'portal',
        title: true
      },
      {
        id: 'cancel-button-portal-id',
        kind: 'portal'
      },
      {
        id: 'create-button-portal-id',
        kind: 'portal'
      }
    ]
    const tooltip = '' //{ text: msgs.get('tooltip.text.createCluster', locale), link: TOOLTIP_LINKS.CREATE_CLUSTER }
    const title =
      selectedAppName || msgs.get(secondaryHeaderProps.title, locale)
    const label = secondaryHeaderProps.label
      ? msgs.get(secondaryHeaderProps.label, locale)
      : undefined
    const labelColor = secondaryHeaderProps.labelColor
      ? msgs.get(secondaryHeaderProps.labelColor, locale)
      : undefined
    this.props.updateSecondaryHeader(
      title,
      secondaryHeaderProps.tabs,
      this.getBreadcrumbs(),
      portals,
      null,
      tooltip,
      label,
      labelColor
    )

    canCreateActionAllNamespaces(
      'applicationset',
      'create',
      'argoproj.io'
    ).then(response => {
      const hasPermissions = _.get(response, 'data.userAccessAnyNamespaces')
      this.setState({ hasPermissions })
    })
  }

  componentDidUpdate() {
    const { mutateStatus, cleanReqStatus, history } = this.props
    if (mutateStatus && mutateStatus === 'DONE') {
      setTimeout(() => {
        if (cleanReqStatus) {
          this.props.cleanReqStatus()
        }
        // redirect to table page
        // known limitation that the appset wouldn't be able to find the clusterName in create
        history.push(`${config.contextPath}`)
      }, 3000)
    }
  }

  render() {
    const editApplication = this.getEditApplication()
    if (editApplication) {
      // if editing an existing app, grab it first
      const { selectedAppName, selectedAppNamespace } = editApplication
      return (
        <Query
          query={getArgoApplication}
          variables={{
            name: selectedAppName,
            namespace: selectedAppNamespace,
            apiversion: 'argoproj.io/v1alpha1'
          }}
        >
          {result => {
            const { loading } = result
            const { data = {} } = result
            const { application } = data
            const errored = application ? false : true
            const error = application ? null : result.error
            if (!loading && error) {
              const errorName = result.error.graphQLErrors[0].name
                ? result.error.graphQLErrors[0].name
                : error.name
              error.name = errorName
            }

            const fetchControl = {
              resources: getApplicationResources(application),
              isLoaded: !loading,
              isFailed: errored,
              error: error
            }

            return this.renderEditor(fetchControl)
          }}
        </Query>
      )
    }
    return this.renderEditor()
  }

  renderEditor(fetchControl) {
    const editApplication = this.getEditApplication()
    const { locale } = this.context
    const { hasPermissions } = this.state
    const { mutateStatus, mutateErrorMsgs } = this.props
    const createControl = {
      hasPermissions,
      createResource: this.handleCreate.bind(this),
      cancelCreate: this.handleCancel.bind(this),
      creationStatus: mutateStatus,
      creationMsg: mutateErrorMsgs
    }
    const i18n = (key, arg2) => {
      return msgs.get(key, arg2, locale)
    }
    return (
      controlData && (
        <PageSection
          className={classNames({ editApplication, 'pf-c-content': true })}
          style={{ padding: 0 }}
          variant="light"
        >
          <TemplateEditor
            type={'argo'}
            title={msgs.get('creation.argo.yaml', locale)}
            template={createTemplate}
            controlData={controlData}
            monacoEditor={<MonacoEditor />}
            portals={Portals}
            fetchControl={fetchControl}
            createControl={createControl}
            i18n={i18n}
          />
        </PageSection>
      )
    )
  }

  handleCreate = resourceJSON => {
    if (resourceJSON) {
      const { handleCreateApplication, handleUpdateApplication } = this.props
      const editApplication = this.getEditApplication()
      if (editApplication) {
        resourceJSON.createResources.push({
          deleteLinks: [...resourceJSON.deleteResources]
        })
        handleUpdateApplication(resourceJSON.createResources)
      } else {
        handleCreateApplication(resourceJSON.createResources)
      }
      const map = _.keyBy(resourceJSON.createResources, 'kind')
      this.applicationNamespace = _.get(
        map,
        'ApplicationSet.metadata.namespace'
      )
      this.applicationName =
        _.get(map, 'ApplicationSet.metadata.name') + '-in-cluster'
      this.apiVersion = _.get(map, 'ApplicationSet.apiVersion')
    }
  };

  handleCancel = () => {
    const { location, history } = this.props
    const editApplication = this.getEditApplication()
    if (location.state && location.state.cancelBack) {
      // Came from the "Create application" button, or "Edit application" table action; go back
      history.goBack()
    } else if (editApplication && location.state && location.state.tabChange) {
      // Came from changing tabs from "Overview" to "Editor"; change tabs back
      history.replace(location.state.tabChange)
    } else {
      // Otherwise, navigate to applications
      history.push(config.contextPath)
    }
  };
}

ArgoCreationPage.contextTypes = {
  locale: PropTypes.string
}

const mapDispatchToProps = dispatch => {
  return {
    cleanReqStatus: () => dispatch(clearCreateStatus()),
    handleCreateApplication: json => dispatch(createApplication(json)),
    handleUpdateApplication: json => dispatch(updateApplication(json)),
    updateSecondaryHeader: (
      title,
      tabs,
      breadcrumbItems,
      ports,
      actions,
      tooltip,
      label,
      labelColor
    ) =>
      dispatch(
        updateSecondaryHeader(
          title,
          tabs,
          breadcrumbItems,
          ports,
          actions,
          tooltip,
          label,
          labelColor
        )
      )
  }
}

const mapStateToProps = state => {
  const { applicationPageResources } = state
  const { mutateStatus, mutateErrorMsgs } = applicationPageResources || {}
  return {
    mutateStatus,
    mutateErrorMsgs
  }
}

export default withRouter(
  connect(mapStateToProps, mapDispatchToProps)(ArgoCreationPage)
)
