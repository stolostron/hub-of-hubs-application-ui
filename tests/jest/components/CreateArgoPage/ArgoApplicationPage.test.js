// Copyright (c) 2021 Red Hat, Inc.
// Copyright Contributors to the Open Cluster Management project

const React = require("react");

import renderer from "react-test-renderer";
import * as reducers from "../../../../src-web/reducers";

import { createStore, combineReducers, applyMiddleware, compose } from "redux";
import thunkMiddleware from "redux-thunk";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";

import ArgoCreationPage from "../../../../src-web/components/ArgoCreationPage/ArgoCreationPage";

const middleware = [thunkMiddleware];
const preloadedState = window.__PRELOADED_STATE__;
const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

const store = createStore(
  combineReducers(reducers),
  preloadedState,
  composeEnhancers(applyMiddleware(...middleware))
);

const locale = "en-US";
describe("ArgoCreationPage creating application set", () => {
  it("ArgoCreationPage renders correctly when creating application", () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <Provider store={store}>
            <ArgoCreationPage
              params={params}
              serverProps={{
                title: "Red Hat Advanced Cluster Management for Kubernetes",
                context: { locale: "en-US" }
              }}
              secondaryHeaderProps={{ title: "title: 'argo.create.title' " }}
            />
          </Provider>
        </BrowserRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

const params = {
  name: "appset1",
  namespace: "default"
};
