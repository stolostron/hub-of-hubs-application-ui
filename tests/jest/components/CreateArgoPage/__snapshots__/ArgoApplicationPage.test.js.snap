// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArgoCreationPage creating application set ArgoCreationPage renders correctly when creating application 1`] = `
<section
  className="pf-c-page__main-section pf-m-light pf-c-content"
  style={
    Object {
      "padding": 0,
    }
  }
>
  <div
    className="temptifly showWizard"
  >
    <div
      className="creation-view-split-container"
    >
      <div
        className="SplitPane  vertical "
        style={
          Object {
            "MozUserSelect": "text",
            "WebkitUserSelect": "text",
            "display": "flex",
            "flex": 1,
            "flexDirection": "row",
            "height": "100%",
            "left": 0,
            "msUserSelect": "text",
            "outline": "none",
            "overflow": "hidden auto",
            "position": "absolute",
            "right": 0,
            "userSelect": "text",
          }
        }
      >
        <div
          className="Pane vertical Pane1  "
          style={
            Object {
              "flex": "none",
              "outline": "none",
              "position": "relative",
              "width": "100%",
            }
          }
        >
          <div
            className="creation-view-controls-container"
          >
            <div
              className="pf-c-wizard"
              style={
                Object {
                  "height": "100%",
                }
              }
            >
              <button
                aria-expanded={false}
                aria-label="Wizard Toggle"
                className="pf-c-wizard__toggle"
                onClick={[Function]}
              >
                <ol
                  className="pf-c-wizard__toggle-list"
                >
                  <li
                    className="pf-c-wizard__toggle-list-item"
                  >
                    <span
                      className="pf-c-wizard__toggle-num"
                    >
                      1
                    </span>
                     
                    <div
                      className="tf--finish-step-button"
                    >
                      <div>
                        General
                      </div>
                    </div>
                  </li>
                </ol>
                <span
                  className="pf-c-wizard__toggle-icon"
                >
                  <svg
                    aria-hidden="true"
                    aria-labelledby={null}
                    fill="currentColor"
                    height="1em"
                    role="img"
                    style={
                      Object {
                        "verticalAlign": "-0.125em",
                      }
                    }
                    viewBox="0 0 320 512"
                    width="1em"
                  >
                    <path
                      d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                    />
                  </svg>
                </span>
              </button>
              <div
                className="pf-c-wizard__outer-wrap"
              >
                <div
                  className="pf-c-wizard__inner-wrap"
                >
                  <nav
                    aria-label="Create wizard steps"
                    aria-labelledby={null}
                    className="pf-c-wizard__nav"
                  >
                    <ol
                      className="pf-c-wizard__nav-list"
                    >
                      <li
                        className="pf-c-wizard__nav-item"
                      >
                        <button
                          aria-current="page"
                          aria-disabled={null}
                          className="pf-c-wizard__nav-link pf-m-current"
                          disabled={false}
                          onClick={[Function]}
                        >
                          <div
                            className="tf--finish-step-button"
                          >
                            <div>
                              General
                            </div>
                          </div>
                        </button>
                      </li>
                      <li
                        className="pf-c-wizard__nav-item"
                      >
                        <button
                          aria-current={false}
                          aria-disabled={null}
                          className="pf-c-wizard__nav-link"
                          disabled={false}
                          onClick={[Function]}
                        >
                          <div
                            className="tf--finish-step-button"
                          >
                            <div>
                              Template
                            </div>
                          </div>
                        </button>
                      </li>
                      <li
                        className="pf-c-wizard__nav-item"
                      >
                        <button
                          aria-current={false}
                          aria-disabled={null}
                          className="pf-c-wizard__nav-link"
                          disabled={false}
                          onClick={[Function]}
                        >
                          <div
                            className="tf--finish-step-button"
                          >
                            <div>
                              Sync policy
                            </div>
                          </div>
                        </button>
                      </li>
                      <li
                        className="pf-c-wizard__nav-item"
                      >
                        <button
                          aria-current={false}
                          aria-disabled={null}
                          className="pf-c-wizard__nav-link"
                          disabled={false}
                          onClick={[Function]}
                        >
                          <div
                            className="tf--finish-step-button"
                          >
                            <div>
                              Placement
                            </div>
                          </div>
                        </button>
                      </li>
                      <li
                        className="pf-c-wizard__nav-item"
                      >
                        <button
                          aria-current={false}
                          aria-disabled={null}
                          className="pf-c-wizard__nav-link"
                          disabled={false}
                          onClick={[Function]}
                        >
                          Review
                        </button>
                      </li>
                    </ol>
                  </nav>
                  <div
                    aria-label="Create wizard content"
                    aria-labelledby={null}
                    className="pf-c-wizard__main"
                  >
                    <div
                      className="pf-c-wizard__main-body"
                    >
                      <div
                        className="creation-view-controls pf-c-form"
                      >
                        <h2>
                          General
                        </h2>
                        <div
                          className="creation-view-controls-section"
                        >
                          <div
                            className="creation-view-controls-textbox"
                            style={
                              Object {
                                "display": "",
                              }
                            }
                          >
                            <div
                              className="pf-c-form__group"
                              id="appSetName-label"
                            >
                              <div
                                className="pf-c-form__group-label"
                              >
                                <label
                                  className="pf-c-form__label"
                                  htmlFor="appSetName"
                                >
                                  <span
                                    className="pf-c-form__label-text"
                                  >
                                    Argo CD ApplicationSet name
                                  </span>
                                  <span
                                    aria-hidden="true"
                                    className="pf-c-form__label-required"
                                  >
                                     
                                    *
                                  </span>
                                </label>
                                 
                                <button
                                  aria-label="More info"
                                  className="pf-c-form__group-label-help"
                                  id="appSetName-label-help-button"
                                  onClick={[Function]}
                                >
                                  <svg
                                    aria-hidden={true}
                                    aria-labelledby={null}
                                    fill="currentColor"
                                    height="1em"
                                    role="img"
                                    style={null}
                                    viewBox="0 0 1024 1024"
                                    width="1em"
                                  >
                                    <path
                                      d="M521.3,576 C627.5,576 713.7,502 713.7,413.7 C713.7,325.4 627.6,253.6 521.3,253.6 C366,253.6 334.5,337.7 329.2,407.2 C329.2,414.3 335.2,416 343.5,416 L445,416 C450.5,416 458,415.5 460.8,406.5 C460.8,362.6 582.9,357.1 582.9,413.6 C582.9,441.9 556.2,470.9 521.3,473 C486.4,475.1 447.3,479.8 447.3,521.7 L447.3,553.8 C447.3,570.8 456.1,576 472,576 C487.9,576 521.3,576 521.3,576 M575.3,751.3 L575.3,655.3 C575.313862,651.055109 573.620137,646.982962 570.6,644 C567.638831,640.947672 563.552355,639.247987 559.3,639.29884 L463.3,639.29884 C459.055109,639.286138 454.982962,640.979863 452,644 C448.947672,646.961169 447.247987,651.047645 447.29884,655.3 L447.29884,751.3 C447.286138,755.544891 448.979863,759.617038 452,762.6 C454.961169,765.652328 459.047645,767.352013 463.3,767.30116 L559.3,767.30116 C563.544891,767.313862 567.617038,765.620137 570.6,762.6 C573.659349,759.643612 575.360354,755.553963 575.3,751.3 M512,896 C300.2,896 128,723.9 128,512 C128,300.3 300.2,128 512,128 C723.8,128 896,300.2 896,512 C896,723.8 723.7,896 512,896 M512.1,0 C229.7,0 0,229.8 0,512 C0,794.2 229.8,1024 512.1,1024 C794.4,1024 1024,794.3 1024,512 C1024,229.7 794.4,0 512.1,0"
                                    />
                                  </svg>
                                </button>
                              </div>
                              <div
                                className="pf-c-form__group-control"
                              >
                                <input
                                  aria-invalid={false}
                                  aria-label={null}
                                  className="pf-c-form-control"
                                  data-ouia-component-id="OUIA-Generated-TextInputBase-1"
                                  data-ouia-component-type="PF4/TextInput"
                                  data-ouia-safe={true}
                                  data-testid="text-appSetName"
                                  disabled={false}
                                  id="appSetName"
                                  onBlur={[Function]}
                                  onChange={[Function]}
                                  onFocus={[Function]}
                                  placeholder="Enter the application set name"
                                  readOnly={false}
                                  required={false}
                                  spellCheck="false"
                                  type="text"
                                  value=""
                                />
                                
                              </div>
                            </div>
                          </div>
                          <div
                            className="creation-view-controls-combobox"
                          >
                            <div
                              className="pf-c-form__group"
                              id="argoServer-label"
                            >
                              <div
                                className="pf-c-form__group-label"
                              >
                                <label
                                  className="pf-c-form__label"
                                  htmlFor="argoServer"
                                >
                                  <span
                                    className="pf-c-form__label-text"
                                  >
                                    Argo server
                                  </span>
                                  <span
                                    aria-hidden="true"
                                    className="pf-c-form__label-required"
                                  >
                                     
                                    *
                                  </span>
                                </label>
                                 
                                <button
                                  aria-label="More info"
                                  className="pf-c-form__group-label-help"
                                  id="argoServer-label-help-button"
                                  onClick={[Function]}
                                >
                                  <svg
                                    aria-hidden={true}
                                    aria-labelledby={null}
                                    fill="currentColor"
                                    height="1em"
                                    role="img"
                                    style={null}
                                    viewBox="0 0 1024 1024"
                                    width="1em"
                                  >
                                    <path
                                      d="M521.3,576 C627.5,576 713.7,502 713.7,413.7 C713.7,325.4 627.6,253.6 521.3,253.6 C366,253.6 334.5,337.7 329.2,407.2 C329.2,414.3 335.2,416 343.5,416 L445,416 C450.5,416 458,415.5 460.8,406.5 C460.8,362.6 582.9,357.1 582.9,413.6 C582.9,441.9 556.2,470.9 521.3,473 C486.4,475.1 447.3,479.8 447.3,521.7 L447.3,553.8 C447.3,570.8 456.1,576 472,576 C487.9,576 521.3,576 521.3,576 M575.3,751.3 L575.3,655.3 C575.313862,651.055109 573.620137,646.982962 570.6,644 C567.638831,640.947672 563.552355,639.247987 559.3,639.29884 L463.3,639.29884 C459.055109,639.286138 454.982962,640.979863 452,644 C448.947672,646.961169 447.247987,651.047645 447.29884,655.3 L447.29884,751.3 C447.286138,755.544891 448.979863,759.617038 452,762.6 C454.961169,765.652328 459.047645,767.352013 463.3,767.30116 L559.3,767.30116 C563.544891,767.313862 567.617038,765.620137 570.6,762.6 C573.659349,759.643612 575.360354,755.553963 575.3,751.3 M512,896 C300.2,896 128,723.9 128,512 C128,300.3 300.2,128 512,128 C723.8,128 896,300.2 896,512 C896,723.8 723.7,896 512,896 M512.1,0 C229.7,0 0,229.8 0,512 C0,794.2 229.8,1024 512.1,1024 C794.4,1024 1024,794.3 1024,512 C1024,229.7 794.4,0 512.1,0"
                                    />
                                  </svg>
                                </button>
                              </div>
                              <div
                                className="pf-c-form__group-control"
                              >
                                <div
                                  className="creation-view-controls-singleselect-loading  pf-c-form-control"
                                >
                                  <span
                                    aria-valuetext="Loading..."
                                    className="pf-c-spinner pf-m-md"
                                    role="progressbar"
                                  >
                                    <span
                                      className="pf-c-spinner__clipper"
                                    />
                                    <span
                                      className="pf-c-spinner__lead-ball"
                                    />
                                    <span
                                      className="pf-c-spinner__tail-ball"
                                    />
                                  </span>
                                  <div />
                                </div>
                                
                              </div>
                            </div>
                          </div>
                          <div
                            className="creation-view-controls-add-value-container bottom-right"
                          >
                            <div
                              className="creation-view-controls-add-button"
                              id="clusterSetLink"
                              onClick={[Function]}
                              onKeyPress={[Function]}
                              role="button"
                              tabIndex="0"
                            >
                              Add cluster sets
                              <svg
                                aria-hidden={true}
                                fill="currentColor"
                                height="1em"
                                icon="opennewtab"
                                role="presentation"
                                style={
                                  Object {
                                    "verticalAlign": "-0.125em",
                                  }
                                }
                                viewBox="0 0 1024 1024"
                                width="1em"
                              >
                                <g
                                  fillRule="evenodd"
                                  stroke="none"
                                  strokeWidth="1"
                                >
                                  <path
                                    d="M576,320 L896,320 L896,192 L576,192 L576,320 Z M128,320 L448,320 L448,192 L128,192 L128,320 Z M930,64 L896,64 L128,64 L94,64 C42.085,64 0,106.085 0,158 L0,192 L0,832 L0,866 C0,917.915 42.085,960 94,960 L128,960 L488,960 C501.255,960 512,949.255 512,936 L512,856 C512,842.745 501.255,832 488,832 L140,832 C133.373,832 128,826.627 128,820 L128,448 L896,448 L896,552 C896,565.255 906.745,576 920,576 L1000,576 C1013.255,576 1024,565.255 1024,552 L1024,158 C1024,106.085 981.915,64 930,64 L930,64 Z"
                                  />
                                  <path
                                    d="M968,784 L848,784 L848,664 C848,650.7 837.3,640 824,640 L776,640 C762.7,640 752,650.7 752,664 L752,784 L632,784 C618.7,784 608,794.7 608,808 L608,856 C608,869.3 618.7,880 632,880 L752,880 L752,1000 C752,1013.3 762.7,1024 776,1024 L824,1024 C837.3,1024 848,1013.3 848,1000 L848,880 L968,880 C981.3,880 992,869.3 992,856 L992,808 C992,794.7 981.3,784 968,784"
                                  />
                                </g>
                              </svg>
                            </div>
                          </div>
                          <div
                            className="creation-view-controls-combobox"
                          >
                            <div
                              className="pf-c-form__group"
                              id="requeueTime-label"
                            >
                              <div
                                className="pf-c-form__group-label"
                              >
                                <label
                                  className="pf-c-form__label"
                                  htmlFor="requeueTime"
                                >
                                  <span
                                    className="pf-c-form__label-text"
                                  >
                                    Requeue time
                                  </span>
                                  <span
                                    aria-hidden="true"
                                    className="pf-c-form__label-required"
                                  >
                                     
                                    *
                                  </span>
                                </label>
                                 
                                <button
                                  aria-label="More info"
                                  className="pf-c-form__group-label-help"
                                  id="requeueTime-label-help-button"
                                  onClick={[Function]}
                                >
                                  <svg
                                    aria-hidden={true}
                                    aria-labelledby={null}
                                    fill="currentColor"
                                    height="1em"
                                    role="img"
                                    style={null}
                                    viewBox="0 0 1024 1024"
                                    width="1em"
                                  >
                                    <path
                                      d="M521.3,576 C627.5,576 713.7,502 713.7,413.7 C713.7,325.4 627.6,253.6 521.3,253.6 C366,253.6 334.5,337.7 329.2,407.2 C329.2,414.3 335.2,416 343.5,416 L445,416 C450.5,416 458,415.5 460.8,406.5 C460.8,362.6 582.9,357.1 582.9,413.6 C582.9,441.9 556.2,470.9 521.3,473 C486.4,475.1 447.3,479.8 447.3,521.7 L447.3,553.8 C447.3,570.8 456.1,576 472,576 C487.9,576 521.3,576 521.3,576 M575.3,751.3 L575.3,655.3 C575.313862,651.055109 573.620137,646.982962 570.6,644 C567.638831,640.947672 563.552355,639.247987 559.3,639.29884 L463.3,639.29884 C459.055109,639.286138 454.982962,640.979863 452,644 C448.947672,646.961169 447.247987,651.047645 447.29884,655.3 L447.29884,751.3 C447.286138,755.544891 448.979863,759.617038 452,762.6 C454.961169,765.652328 459.047645,767.352013 463.3,767.30116 L559.3,767.30116 C563.544891,767.313862 567.617038,765.620137 570.6,762.6 C573.659349,759.643612 575.360354,755.553963 575.3,751.3 M512,896 C300.2,896 128,723.9 128,512 C128,300.3 300.2,128 512,128 C723.8,128 896,300.2 896,512 C896,723.8 723.7,896 512,896 M512.1,0 C229.7,0 0,229.8 0,512 C0,794.2 229.8,1024 512.1,1024 C794.4,1024 1024,794.3 1024,512 C1024,229.7 794.4,0 512.1,0"
                                    />
                                  </svg>
                                </button>
                              </div>
                              <div
                                className="pf-c-form__group-control"
                              >
                                <div
                                  id="requeueTime-group"
                                >
                                  <div
                                    aria-label="Choose an item"
                                    className="tf--list-box"
                                    role="listbox"
                                    tabIndex="0"
                                  >
                                    <div
                                      aria-expanded={false}
                                      aria-haspopup="true"
                                      aria-label="Open menu"
                                      className=""
                                      data-toggle="true"
                                      onClick={[Function]}
                                      onKeyPress={[Function]}
                                      role="button"
                                      tabIndex="0"
                                      type="button"
                                    >
                                      <div
                                        className="pf-c-form-control input"
                                      >
                                        <input
                                          aria-controls="requeueTime-Requeue time-180"
                                          aria-expanded="true"
                                          aria-label="ListBox input field"
                                          autoComplete="off"
                                          className="pf-c-combo-control"
                                          data-testid="combo-requeueTime"
                                          id="requeueTime"
                                          onBlur={[Function]}
                                          onChange={[Function]}
                                          onFocus={[Function]}
                                          onKeyDown={[Function]}
                                          onKeyUp={[Function]}
                                          placeholder="Enter or select a placement name"
                                          role="combobox"
                                          spellCheck="false"
                                          value="180"
                                        />
                                      </div>
                                      <div
                                        className="tf--list-box__selection"
                                        onClick={[Function]}
                                        onKeyPress={[Function]}
                                        role="button"
                                        style={
                                          Object {
                                            "color": "#6a6e73",
                                          }
                                        }
                                        tabIndex="0"
                                        title="Clear selected item"
                                      >
                                        <svg
                                          aria-hidden={true}
                                          aria-labelledby={null}
                                          fill="currentColor"
                                          height="1em"
                                          role="img"
                                          style={
                                            Object {
                                              "verticalAlign": "-0.125em",
                                            }
                                          }
                                          viewBox="0 0 512 512"
                                          width="1em"
                                        >
                                          <path
                                            d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm121.6 313.1c4.7 4.7 4.7 12.3 0 17L338 377.6c-4.7 4.7-12.3 4.7-17 0L256 312l-65.1 65.6c-4.7 4.7-12.3 4.7-17 0L134.4 338c-4.7-4.7-4.7-12.3 0-17l65.6-65-65.6-65.1c-4.7-4.7-4.7-12.3 0-17l39.6-39.6c4.7-4.7 12.3-4.7 17 0l65 65.7 65.1-65.6c4.7-4.7 12.3-4.7 17 0l39.6 39.6c4.7 4.7 4.7 12.3 0 17L312 256l65.6 65.1z"
                                          />
                                        </svg>
                                      </div>
                                      <div
                                        className="tf--list-box__menu-icon"
                                        onClick={[Function]}
                                        onKeyPress={[Function]}
                                        role="button"
                                        tabIndex="0"
                                      >
                                        <svg
                                          alt="Open menu"
                                          aria-label="Open menu"
                                          fillRule="evenodd"
                                          height="5"
                                          role="img"
                                          viewBox="0 0 10 5"
                                          width="10"
                                        >
                                          <title>
                                            Close menu
                                          </title>
                                          <path
                                            d="M0 0l5 4.998L10 0z"
                                          />
                                        </svg>
                                      </div>
                                    </div>
                                  </div>
                                </div>
                                
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <footer
                  className="pf-c-wizard__footer"
                >
                  <button
                    aria-disabled={false}
                    aria-label={null}
                    className="pf-c-button pf-m-primary pf-m-progress"
                    data-ouia-component-id="OUIA-Generated-Button-primary-1"
                    data-ouia-component-type="PF4/Button"
                    data-ouia-safe={true}
                    disabled={false}
                    onClick={[Function]}
                    role={null}
                    type="button"
                  >
                    Next
                  </button>
                  <button
                    aria-disabled={true}
                    aria-label={null}
                    className="pf-c-button pf-m-secondary pf-m-aria-disabled"
                    data-ouia-component-id="OUIA-Generated-Button-secondary-1"
                    data-ouia-component-type="PF4/Button"
                    data-ouia-safe={true}
                    disabled={false}
                    onClick={[Function]}
                    onKeyPress={[Function]}
                    role={null}
                    tabIndex={null}
                    type="button"
                  >
                    Back
                  </button>
                  <button
                    aria-disabled={false}
                    aria-label={null}
                    className="pf-c-button pf-m-link"
                    data-ouia-component-id="OUIA-Generated-Button-link-1"
                    data-ouia-component-type="PF4/Button"
                    data-ouia-safe={true}
                    disabled={false}
                    onClick={[Function]}
                    role={null}
                    type="button"
                  >
                    Cancel
                  </button>
                </footer>
              </div>
            </div>
          </div>
        </div>
        <span
          className="Resizer vertical "
          onClick={[Function]}
          onDoubleClick={[Function]}
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
          onTouchStart={[Function]}
          role="presentation"
          style={Object {}}
        />
        <div
          className="Pane vertical Pane2  "
          style={
            Object {
              "flex": 1,
              "outline": "none",
              "position": "relative",
            }
          }
        />
      </div>
    </div>
  </div>
</section>
`;
